src/gpu/optixPrograms.cu

```diff
@@ -141,69 +69,91 @@ extern "C" __global__ void __raygen__()
 		// Velocities are in the local frame. Need to operate on rays in local frame.
 		// Ray time offsets are in milliseconds, velocities are in unit per seconds.
 		// In order to not lose numerical precision, first multiply values and then convert to proper unit.
-		ray = Mat3x4f::TRS((ctx.rayTimeOffsets[rayIdx] * ctx.sensorLinearVelocityXYZ) * 0.001f,
-		                   (ctx.rayTimeOffsets[rayIdx] * (ctx.sensorAngularVelocityRPY * toDeg)) * 0.001f) *
+		ray = Mat3x4f::TRS((ctx.rayTimeOffsetsMs[rayIdx] * ctx.sensorLinearVelocityXYZ) * 0.001f,
+		                   (ctx.rayTimeOffsetsMs[rayIdx] * (ctx.sensorAngularVelocityRPY * toDeg)) * 0.001f) *
 		      rayLocal;
 		// Back to the global frame.
 		ray = ctx.rayOriginToWorld * ray;
 	}
 
-	Vec3f origin = ray * Vec3f{0, 0, 0};
-	Vec3f dir = ray * Vec3f{0, 0, 1} - origin;
 	float maxRange = ctx.rayRangesCount == 1 ? ctx.rayRanges[0].y() : ctx.rayRanges[rayIdx].y();
 
-	unsigned int flags = OPTIX_RAY_FLAG_DISABLE_ANYHIT;
-	Vec3fPayload originPayload = encodePayloadVec3f(origin);
-	optixTrace(ctx.scene, origin, dir, 0.0f, maxRange, 0.0f, OptixVisibilityMask(255), flags, 0, 1, 0, originPayload.p0,
-	           originPayload.p1, originPayload.p2);
+	shootSamplingRay(ray, maxRange, 0); // Shoot primary ray
+	if (ctx.beamHalfDivergence > 0.0f) {
```

Comment:
I would make multi-sampling dependent on the expected result or raytrace setting, not on the lidar configuration. To be resolved in the next PR anyway.

Reply:
Why would you do so? The only reason to set `beamDivergence` is when you're interested in multi-return output. This way, `beamHalfDivergence` becomes the only source of truth whether we should perform multi-sampling.

This is different from LiDAR velocity and applying distortion, because LiDAR velocity is used also to compute `RELATIVE_VELOCITY`, so we need additional `ctx.doApplyDistortion` there.

Reply:
Resolved offline